Support programmatic `multipart/form-data` responses

Sgitario opened the pull request:
With these changes, users can also manually append the parts of the form using the class `MultipartFormDataOutput` as:

```java
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

import org.jboss.resteasy.reactive.server.core.multipart.MultipartFormDataOutput;

@Path("multipart")
public class Endpoint {

    @GET
    @Produces(MediaType.MULTIPART_FORM_DATA)
    @Path("file")
    public MultipartFormDataOutput getFile() {
        MultipartFormDataOutput form = new MultipartFormDataOutput();
        form.addFormData("person", new Person("John"), MediaType.APPLICATION_JSON_TYPE);
        form.addFormData("status", "a status", MediaType.TEXT_PLAIN_TYPE)
                .getHeaders().putSingle("extra-header", "extra-value");
        return form;
    }
}
```

Fix https://github.com/quarkusio/quarkus/issues/28631

This last approach allows you adding extra headers to the output part.

Sgitario commented:
cc @geoand @knutwannheden

geoand commented on independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/multipart/MultipartFormDataOutput.java:
```diff
@@ -0,0 +1,25 @@
+import javax.ws.rs.core.MediaType;
+
+public class MultipartFormDataOutput {
+    private Map<String, PartItem> parts = new HashMap<>();
```
This can be `final`

geoand reviewed (approved):
Very nice!

I just added one tiny comment

Sgitario commented:
> MultipartFormDataOutput



> It would be nice if the individual parts could also declare headers as in the classic implementation I mentioned. But I can also create a separate feature request for that.

I also did this as part of this pull request.

geoand commented:
@Sgitario CI doesn't seem to like this PR :)

Sgitario commented:
> @Sgitario CI doesn't seem to like this PR :)

Indeed. I'm having another look

knutwannheden commented:
> I also did this as part of this pull request.

Excellent! I must take a closer look then.

Sgitario committed:
Support programmatic `multipart/form-data` responses

With these changes, users can also manually append the parts of the form using the class `MultipartFormDataOutput` as:

```java
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

import org.jboss.resteasy.reactive.server.core.multipart.MultipartFormDataOutput;

@Path("multipart")
public class Endpoint {

    @GET
    @Produces(MediaType.MULTIPART_FORM_DATA)
    @Path("file")
    public MultipartFormDataOutput getFile() {
        MultipartFormDataOutput form = new MultipartFormDataOutput();
        form.addFormData("person", new Person("John"), MediaType.APPLICATION_JSON_TYPE);
        form.addFormData("status", "a status", MediaType.TEXT_PLAIN_TYPE)
                .getHeaders().putSingle("extra-header", "extra-value");
        return form;
    }
}
```

Fix https://github.com/quarkusio/quarkus/issues/28631

This last approach allows you adding extra headers to the output part.

Sgitario commented:
PR updated with the following changes to address the CI failures:
- This was a mistake on my side: when the parts had no filename, it was printing "null".
- Because the parts are not ordered any longer (they are stored in a map), when serializing an entity (using JSON), the JsonB writer was closing the output stream, and hence the rest of the parts were not written. To fix this issue, I'm using a byte array output stream for the entity writers.

knutwannheden commented:
I am wondering if possibly the parts should maintain insertion order. I will try to play with it and check if we have any issues with it. Thanks!
